### DISCUSSION TITLE
How to use self compiled static ffmpeg.

### DISCUSSION THREAD

**Discussion by zdyshine** [python-ffmpegio/python-ffmpegio#32] [Q&A]
Great Work, and How to use self compiled static ffmpeg?

**Comment by SohamTilekar** [comment]
> Great Work, and How to use self compiled static ffmpeg?

?? @zdyshine

**Comment by SohamTilekar** [comment]
You will also need the ffprobe, no maters it is self compiled or downlod from online. use the `ffmpegio.set_path()` function to set there path.

Or

Put both ffmpeg & ffprobe binay in the below ffmpeg folder. On windos or put binary in the system or user path variables.There name should be same ffmpeg.exe, ffperobe.exe or ffmpeg, ffprobe.

`%PROGRAMFILES%\ffmpeg`

`%PROGRAMFILES(X86)%\ffmpeg`

`%USERPROFILE%\ffmpeg`

`%APPDATA%\ffmpeg`

`%APPDATA%\programs\ffmpeg`

`%LOCALAPPDATA%\ffmpeg`

`%LOCALAPPDATA%\programs\ffmpeg`

On linux put them in the path variable.

**Comment by zdyshine** [comment]
> You will also need the ffprobe, no maters it is self compiled or downlod from online. use the `ffmpegio.set_path()` function to set there path.
> 
> Or
> 
> Put both ffmpeg & ffprobe binay in the below ffmpeg folder. On windos or put binary in the system or user path variables.There name should be same ffmpeg.exe, ffperobe.exe or ffmpeg, ffprobe.
> 
> `%PROGRAMFILES%\ffmpeg`
> 
> `%PROGRAMFILES(X86)%\ffmpeg`
> 
> `%USERPROFILE%\ffmpeg`
> 
> `%APPDATA%\ffmpeg`
> 
> `%APPDATA%\programs\ffmpeg`
> 
> `%LOCALAPPDATA%\ffmpeg`
> 
> `%LOCALAPPDATA%\programs\ffmpeg`
> 
> On linux put them in the path variable.

Thank you for your reply！
I change https://github.com/python-ffmpegio/python-ffmpegio/blob/main/src/ffmpegio/path.py line 30:  FFMPEG_BIN = None. to FFMPEG_BIN : /test/ffmpeg. 
Is this specified method feasible?
When I change, run the code is error:

from ffmpegio import ffmpeg
ffmpeg("-hide_banner -y -loglevel error -i /test/deg.png -vf 'scale=ceil(iw/4):ceil(ih/4):flags=lanczos+accurate_rnd+full_chroma_int:sws_dither=none:param0=5' -c:v libsvtav1 -qp 31 -preset 5 1_4x_qp31.avif ")
TypeError: 'module' object is not callable

I'm not sure if it's a code modification issue or a command issue

**Comment by SohamTilekar** [comment]
Try noto modify the path.py directly use the `ffmpegio.set_path()` function at start of your code.
Provide whole traceback.
Python version
OS


Hav you read the Docs.

**Comment by SohamTilekar** [comment]
There is a ffmpeg function in the ffmpegio. Use the 
```python
import ffmpegio.ffmpeg as fio

fio.ffmpeg()
```

Sory for me being rood do you know how the python works.
How to code.
Which text editor you are using.

**Comment by SohamTilekar** [comment]
@tikuma-lsuhsc I thing he (@zdyshine) caught a bug or outdated doc or reame.md file. I can't test the code. My laptop charger burner could you test it.

**Comment by zdyshine** [comment]
Thanks all. I use ffmpegio.set_path() and solved the problem~
For this code, 
ffmpegio.transcode('/test/deg.png', './output.avif', vf='scale=ceil(iw/4):ceil(ih/4):flags=lanczos+accurate_rnd+full_chroma_int:sws_dither=none:param0=5', qp=31, preset=5,**{'c:v':'libsvtav1'})

How can I directly write the output to IO or get numpy output？

**Comment by tikuma-lsuhsc** [comment]
@zdyshine - You got it: `set_path()` is the way to go. For your latest question, you should use `ffmpegio.image.read()` with all your arguments except for the output. If you have a video input (or a sequence of images), use `ffmpegio.video.read()`

Finally, I'll move this thread to the discussions later as you're not raising an issue. But, feel free to ask additional questions on this thread.

**Comment by tikuma-lsuhsc** [comment]
Actually...

> How can I directly write the output to IO or get numpy output？

Do you mean to retrieve the image data as an AVIF-encoded byte sequence or as a raw RGB matrix?

**Comment by zdyshine** [comment]
> Actually...
> 
> > How can I directly write the output to IO or get numpy output？
> 
> Do you mean to retrieve the image data as an AVIF-encoded byte sequence or as a raw RGB matrix?

I want to obtain an RGB matrix and continue with other processing using Python.

**Comment by tikuma-lsuhsc** [comment]
Then, as I said above, use `ffmpegio.image.read`:

```python
F = ffmpegio.image.read('/test/deg.png', vf='scale=ceil(iw/4):ceil(ih/4):flags=lanczos+accurate_rnd+full_chroma_int:sws_dither=none:param0=5')
```

This should output `HxWx3` RGB24 Numpy array.

Note that the encoder options (`qp=31, preset=5,**{'c:v':'libsvtav1'}`) are not used here.

**Reply by zdyshine** [reply]
yeah, I can get a  RGB24 Numpy array buy ffmpegio.image.read. But I have no idea the encoder options (qp=31, preset=5,**{'c:v':'libsvtav1'}) how to use. I have try the  ffmpegio.image.write, ffmpegio.image.filter and ffmpegprocess.run, but couldn't solve it. The given ffmpegprocess.run command is incorrect, can you check it out.

**Reply by zdyshine** [reply]
I see that ffmpegio.transcode can return a matrix. Can you tell me how to set the parameters

**Reply by tikuma-lsuhsc** [reply]
OK, what do you want to do exactly? You originally asked:

> How can I directly write the output to IO or get numpy output？

I assumed "OR" as in how to do each operation. You already have the `transcode` command, and the `ffmpegio.image.read` above does the raw IO. Do you also need to write an AVIF image from a Numpy array?

**Reply by zdyshine** [reply]
I'm sorry for the inconvenience caused to you. Actually, what I ultimately want is an AVIF image, but it can be directly written in IO instead of local, to reduce the time consumption caused by reading and writing.

**Reply by tikuma-lsuhsc** [reply]
Oh, so you do want an AVIF-encoded byte array as an output (I guess as well as the raw array). This you can do. Just use `-` as the output should do the trick.

```python
b = ffmpegio.transcode('/test/deg.png', '-', vf='scale=ceil(iw/4):ceil(ih/4):flags=lanczos+accurate_rnd+full_chroma_int:sws_dither=none:param0=5', qp=31, preset=5,**{'c:v':'libsvtav1'})
```

**Reply by zdyshine** [reply]
Actually, I am doing an image restoration task and need to repair the compressed image of AV1.

**Comment by zdyshine** [comment]
> Oh, so you do want an AVIF-encoded byte array as an output (I guess as well as the raw array). This you can do. Just use `-` as the output should do the trick.
> 
> ```python
> b = ffmpegio.transcode('/test/deg.png', '-', vf='scale=ceil(iw/4):ceil(ih/4):flags=lanczos+accurate_rnd+full_chroma_int:sws_dither=none:param0=5', qp=31, preset=5,**{'c:v':'libsvtav1'})
> ```

I used this command earlier, but I reported an error:

```
Input #0, png_pipe, from '/test/deg.png':
  Duration: N/A, bitrate: N/A
  Stream #0:0: Video: png, rgb24(pc, gbr/unknown/unknown), 2040x1536, 25 fps, 25 tbr, 25 tbn
[AVFormatContext @ 0x5068a40] Unable to choose an output format for 'pipe:'; use a standard extension for the filename or specify the format manually.
[out#0 @ 0x5068900] Error initializing the muxer for pipe:: Invalid argument
Error opening output file -.
Error opening output files: Invalid argument
Traceback (most recent call last):
  File "/test/ffmpeg_dataset.py", line 410, in <module>
    out = ffmpegio.transcode('/test/deg.png', '-', overwrite=True, show_log=True, vf='scale=ceil(iw/4):ceil(ih/4):flags=lanczos+accurate_rnd+full_chroma_int:sws_dither=none:param0=5', qp=31, preset=5,**{'c:v':'libsvtav1'})
  File "/opt/conda/lib/python3.10/site-packages/ffmpegio/transcode.py", line 125, in transcode
    raise FFmpegError(pout.stderr, show_log)
ffmpegio.errors.FFmpegError: FFmpeg failed. Check its log printed above
```

**Reply by tikuma-lsuhsc** [reply]
hum, let me see if I can fix it. (guess it should've stayed in the issues board lol)

**Reply by tikuma-lsuhsc** [reply]
<strike>Oh you know what, I think the issue is that AVIF format requires to seek the file (like writing the header info after filling the payload). Try other format, say `bmp` or `jpg`</strike>

Take that back. The issue is that the ffmpeg doesn't know the format. add `f='avif'` keyword argument (or whatever the container format name.

**Reply by zdyshine** [reply]
After add f='avif', a new error appeared: 

Input #0, png_pipe, from '/test/deg.png':
  Duration: N/A, bitrate: N/A
  Stream #0:0: Video: png, rgb24(pc, gbr/unknown/unknown), 2040x1536, 25 fps, 25 tbr, 25 tbn
Stream mapping:
  Stream #0:0 -> #0:0 (png (native) -> av1 (libsvtav1))
[avif @ 0x384eb80] muxer does not support non seekable output
[out#0/avif @ 0x3851f00] Could not write header (incorrect codec parameters ?): Invalid argument
Error while filtering: Invalid argument

It may be caused by the AVIF format?
I have uploaded a static version of ffmpeg. If you are interested, you can try it out.
https://drive.google.com/drive/folders/1IkfIE6PQyPQahquxu5mlIFdM_ww54M0Y?usp=drive_link

**Reply by tikuma-lsuhsc** [reply]
Yep, that's the error that I crossed out above. Cannot use an output pipe for this format. You can try to use the encoder via `c:v` option and specify a non-seeking container format. I'm not familiar with this myself, but you can try the `data_raw` muxer (https://ffmpeg.org/ffmpeg-formats.html). Of course, the byte array is no longer an AVIF format, so you need to run ffmpeg again to write it to the file later...

What are you planning on doing with this in-memory AVIF byte array?

**Reply by zdyshine** [reply]
Convert  in-memory AVIF byte array to in-memory numpy matrix and use deep learning methods for image restoration. Reading locally will slow down the training speed of deep learning.

**Reply by tikuma-lsuhsc** [reply]
So, are you trying to train a network to manipulate the AVIF-encoded byte data directly to restore (distored) image without using the raw pixel data? Also, to make sure that the network spits out a valid AVIF format data?

Or, are you trying to fix the quantization error introduced by the AVIF encoder?

**Reply by zdyshine** [reply]
You are right, maybe I can try train a network to manipulate the AVIF-encoded byte data directly to restore (distored) image. But how to obtain  in-memory AVIF byte array is still a problem.

**Reply by tikuma-lsuhsc** [reply]
> "maybe I can try..."

If I were you, I'd spend more time setting up my problem so I have a concrete specification of the DL network IO. Don't rely on a random input from a stranger who has no idea what you're trying to do (that's me). (My gut feeling, however, is that it's not what you really want to do but I could be wrong...)

In any case, if you're dead set on training directly on AVIF byte data, you must create AVIF files from original PNG files off line, and use the AVIF files in your training (just use the built-in `open('file.avif','rb')` and `read`).

**Reply by zdyshine** [reply]
Haha, thank you for your patient guidance. I will try more~

**Reply by tikuma-lsuhsc** [reply]
Good luck. Feel free to drop more questions.

**Reply by SohamTilekar** [reply]
Drop more questions but first read the docs.
